latin irregular verbs #36

irina060981 opened the pull request:
1) created latin-verb-irregular.js with filtering by current lemma

2) Added irregularVerb to inflection.constraints

3) Corected getObligatoryMatchList and getOptionalMatchList for creating matching

it need to be used with - https://github.com/alpheios-project/data-models/pull/37

irina060981 commented:
Old version v1
![image](https://user-images.githubusercontent.com/12377640/42132029-1c196d52-7d53-11e8-8407-16e0625b4c3f.png)

kirlat commented on lib/lang/greek/greek-language-dataset.js:
```diff
@@ -13,8 +14,16 @@ import nounFootnotesCSV from '@lib/lang/greek/data/noun/footnotes.csv'
 import pronounFormsCSV from '@lib/lang/greek/data/pronoun/forms.csv'
 import pronounFootnotesCSV from '@lib/lang/greek/data/pronoun/footnotes.csv'
+
+import GroupFeatureType from '@views/lib/group-feature-type.js'
```
Can you please temporarily roll back aliases here and in other places?

kirlat commented on lib/lang/latin/latin-language-dataset.js:
```diff
@@ -368,8 +384,29 @@ export default class LatinLanguageDataset extends LanguageDataset {
 
+  getVerbsIrregularLemma (verbForm) {
+    let res = this.pos.get(Constants.POFS_VERB).types.get(Form).items.map(item => { return { val: item.value, lemma: item.features[Feature.types.hdwd].value, fullForm: item.features[Feature.types.fullForm].value } })
+    let res1 = {}
```
Can we give `res` and `res1` more meaningful names that will reflect what data they contain? This way code will be easier to read. Thanks!

irina060981 replied:
Oh, yes - let's do it.

kirlat commented on lib/lang/latin/latin-language-dataset.js:
```diff
@@ -368,8 +384,29 @@ export default class LatinLanguageDataset extends LanguageDataset {
+  }
+
+  checkIrregularVerb (inflection) {
+    if (inflection.hasFeatureValue(Feature.types.part, Constants.POFS_VERB) && this.getVerbsIrregularLemma(inflection[Feature.types.fullForm])) {
```
Can we simplify this and below to `return (inflection.hasFeatureValue(Feature.types.part, Constants.POFS_VERB) && this.getVerbsIrregularLemma(inflection[Feature.types.fullForm]))`?

irina060981 replied:
Of course

kirlat commented on views/lang/greek/article/greek-article-view.js:
```diff
@@ -0,0 +1,70 @@
+import Form from '@lib/form.js'
+import GreekView from '@views/lang/greek/greek-view.js'
+import GroupFeatureType from '@views/lib/group-feature-type.js'
+import Table from '@views/lib/table.js'
```
Can you please roll back aliases temporarily?

kirlat commented on views/lang/greek/greek-view.js:
```diff
@@ -1,8 +1,8 @@
+import LanguageDatasetFactory from '@lib/language-dataset-factory.js'
+import Table from '@views/lib/table.js'
+import View from '@views/lib/view.js'
+import GroupFeatureType from '@views/lib/group-feature-type.js'
```
Aliases, please, temporarily

kirlat commented on views/lang/greek/numeral/greek-numeral-view.js:
```diff
@@ -0,0 +1,105 @@
+import Table from '@views/lib/table.js'
+
+import GreekView from '@views/lang/greek/greek-view.js'
+import GroupFeatureType from '@views/lib/group-feature-type.js'
```
Aliases

kirlat commented on views/lang/latin/verb/latin-verb-irregular.js:
```diff
@@ -0,0 +1,74 @@
+
+import GroupFeatureType from '@views/lib/group-feature-type'
+import Form from '@lib/form.js'
+import Table from '@views/lib/table'
```
Aliases

kirlat commented on views/lang/latin/verb/latin-verb-irregular.js:
```diff
@@ -0,0 +1,74 @@
+    // default is true
+    for (let lexeme of lexemes) {
+      for (let inflection of lexeme.inflections) {
+        // console.log('************************enabledForLexemes inflection', JSON.stringify(inflection.constraints))
```
Can probably remove a debug statement comment

balmas commented on lib/lang/latin/latin-language-dataset.js:
```diff
@@ -273,17 +273,29 @@ export default class LatinLanguageDataset extends LanguageDataset {
       const item = data[i]
       let lemma = item[0]
       // let principalParts = item[1].split(/_/)
+      let principalParts = item[1]
```
A few notes about the data:
the values in the principal parts field are separated in _ to avoid needing to quote that field in the csv, it's really a list of _ separated values and should be treated as such. that's what why I had before the split on _ line 281.  In our Lemma data model, there is a principalParts field, this is where these values would be found in the lemma of the homonym that would be matched against the table data. I'm not sure we *need* to match on the hdwd and the principal parts but we do want to represent this consistently. So I think we might actually want to instantiate a Lemma object with the hdwd and principal parts from the inflection data.

balmas commented on lib/lang/latin/latin-language-dataset.js:
```diff
@@ -273,17 +273,29 @@ export default class LatinLanguageDataset extends LanguageDataset {
       ]
+
+      if (principalParts) {
+        features.push(this.features.get(Feature.types.hdwd).createFromImporter(principalParts))
```
See note above, the feature value would, I think, be either the Lemma object itself, or a list of the hdwd and the list of prinicipal parts. (e.g. "eo, ire, ivi(ii), itus")

balmas commented on views/lang/latin/verb/latin-verb-irregular.js:
```diff
@@ -0,0 +1,72 @@
+  createTable () {
+    this.table = new Table([this.features.lemmas, this.features.moods, this.features.tenses, this.features.numbers, this.features.persons])
+    let features = this.table.features
+    features.columns = [ this.features.lemmas, this.features.moods ]
```
We need to add a column grouping on voice. Not all of the forms have it, but it if it's present we need to group on it.  E.g. this is why you are getting more values in the cells in the table for "fero" than are there in the V1 version. There are different values for active and passive voice.

balmas commented on views/lang/latin/verb/latin-verb-irregular.js:
```diff
@@ -0,0 +1,72 @@
+    const lemmaValues = this.dataset.getVerbsIrregularLemma(inflectionData.targetWord)
+    let lemmasType = new Feature(Feature.types.hdwd, lemmaValues, LatinVerbIrregularView.languageID)
+
+    this.features = {
```
add voice (see next comment)

balmas commented on lib/lang/latin/latin-language-dataset.js:
```diff
@@ -389,6 +421,15 @@ export default class LatinLanguageDataset extends LanguageDataset {
+        Feature.types.mood,
+        Feature.types.tense,
+        Feature.types.number,
+        Feature.types.person
```
add Feature.types.voice

irina060981 commented:
I have changed grouping (added voice to column), changed saving and filtering by Lemma and not principal parts and fix optional match.
should be used with 
https://github.com/alpheios-project/data-models/pull/37

irina060981 commented:
Tests screenshots

**Ferrant** - irregular verb 

v1
![image](https://user-images.githubusercontent.com/12377640/42208967-d402a278-7ef0-11e8-8fac-56145c442d9e.png)

new version
![image](https://user-images.githubusercontent.com/12377640/42209056-0c9fe690-7ef1-11e8-9552-8ca76b47cf94.png)

irina060981 commented:
**Ferrant** - regular verb

V1
![image](https://user-images.githubusercontent.com/12377640/42209087-25e14234-7ef1-11e8-83ee-245f09c5da07.png)

New version
![image](https://user-images.githubusercontent.com/12377640/42209132-3d8e76cc-7ef1-11e8-82c3-4af52201a9e1.png)

irina060981 commented:
**placito** - Regular verb

V1
![image](https://user-images.githubusercontent.com/12377640/42209203-6fc2aa8c-7ef1-11e8-9854-fede853e1290.png)

New version
![image](https://user-images.githubusercontent.com/12377640/42209234-8138236e-7ef1-11e8-9f44-7c312e489f3e.png)

irina060981 commented:
Sumere - infinite form for irregular verb

V1
![image](https://user-images.githubusercontent.com/12377640/42209482-13097f7c-7ef2-11e8-8f5a-ce9171a86340.png)

New version
![image](https://user-images.githubusercontent.com/12377640/42209502-24b4a8d2-7ef2-11e8-9b7a-7f7f13d0466c.png)

balmas commented on lib/lang/latin/latin-language-dataset.js:
```diff
@@ -279,7 +279,7 @@ export default class LatinLanguageDataset extends LanguageDataset {
       let lemma = item[0]
       // let principalParts = item[1].split(/_/)
-      let principalParts = item[1]
+      // let principalParts = item[1]
```
(And then in the view we would need to set the title of the view using the Lemma object)

irina060981 replied:
Oh, yes. I didn'y understand you correctly before - because it was commented before I started to work on the task, so I didn't think that we need it. I will make these changes.

balmas commented:
Also re 'latus' - as @irina060981 pointed out, this is a verb participle but should match the irregular table for  fero, ferre, tuli, latus, in addition to the verb participle table. 

A simple solution might be to derive LatinVerbParticipleIrregularView from LatinVerbIrregularView and override its getter for partOfSpeech accordingly (to return Constants.POFS_VERB_PARTICIPLE). @kirlat what do you think?

kirlat commented:
@balmas I think that wold be the best solution, simple and flexible.

balmas commented:
ok, thanks @kirlat . @irina060981 can you give that a try?

irina060981 commented:
Yes, I will add it. I didn't guess for this solution.

irina060981 commented:
Inflection table for irregular verbs with additional title
![image](https://user-images.githubusercontent.com/12377640/42273662-215ad16c-7fcd-11e8-848f-3eaab4f58c79.png)

irina060981 commented:
Irregular vreb participles:
![image](https://user-images.githubusercontent.com/12377640/42275353-5fc68a76-7fd3-11e8-961b-9dde16e2f894.png)

irina060981 commented:
@balmas, @kirlat  - all requests are done - waiting for your approval
This request is done with 
https://github.com/alpheios-project/data-models/pull/37
https://github.com/alpheios-project/components/pull/139

balmas commented on views/lang/latin/verb/latin-verb-irregular.js:
```diff
@@ -0,0 +1,80 @@
+    const inflectionsWords = inflectionData.homonym.inflections.map(item => item[Feature.types.word].value)
+    const lemma = this.dataset.verbsIrregularLemmas.filter(item => inflectionsWords.indexOf(item.word) > -1)[0]
+
+    this.additionalTitle = lemma.word + ' - ' + lemma.principalParts
```
@irina060981 please just join the lemma.word to the principal parts with a comma rather than a hyphen and then this is great to merge. Thank you!
